examples/nidcpower_source_dc_voltage_with_multiplexer/measurement.py

```diff
@@ -0,0 +1,171 @@
+
+import click
+import grpc
+import hightime
+import ni_measurementlink_service as nims
+import nidcpower
+import nidcpower.session
+import niswitch.session
+from _helpers import configure_logging, verbosity_option
+
+_NIDCPOWER_WAIT_FOR_EVENT_TIMEOUT_ERROR_CODE = -1074116059
+_NIDCPOWER_TIMEOUT_EXCEEDED_ERROR_CODE = -1074097933
+_NIDCPOWER_TIMEOUT_ERROR_CODES = [
+    _NIDCPOWER_WAIT_FOR_EVENT_TIMEOUT_ERROR_CODE,
+    _NIDCPOWER_TIMEOUT_EXCEEDED_ERROR_CODE,
+]
+
+script_or_exe = sys.executable if getattr(sys, "frozen", False) else __file__
+service_directory = pathlib.Path(script_or_exe).resolve().parent
+measurement_service = nims.MeasurementService(
+    service_config_path=service_directory / "NIDCPowerSourceDCVoltageWithMultiplexer.serviceconfig",
+    version="0.1.0.0",
+    ui_file_paths=[
+        service_directory / "NIDCPowerSourceDCVoltageWithMultiplexer.measui",
+        service_directory / "NIDCPowerSourceDCVoltageWithMultiplexerUI.vi",
+    ],
+)
+
+if TYPE_CHECKING:
+    # The nidcpower Measurement named tuple doesn't support type annotations:
+    # https://github.com/ni/nimi-python/issues/1885
+    class _Measurement(NamedTuple):
+        voltage: float
+        current: float
+        in_compliance: bool
+        channel: str
+
+
+@measurement_service.register_measurement
+@measurement_service.configuration(
+    "pin_name",
+    nims.DataType.Pin,
+    "Pin1",
+    instrument_type=nims.session_management.INSTRUMENT_TYPE_NI_DCPOWER,
+)
+@measurement_service.configuration("voltage_level", nims.DataType.Double, 6.0)
+@measurement_service.configuration("voltage_level_range", nims.DataType.Double, 6.0)
+@measurement_service.configuration("current_limit", nims.DataType.Double, 0.01)
+@measurement_service.configuration("current_limit_range", nims.DataType.Double, 0.01)
+@measurement_service.configuration("source_delay", nims.DataType.Double, 0.0)
+@measurement_service.output("voltage_measurement", nims.DataType.Double)
+@measurement_service.output("current_measurement", nims.DataType.Double)
+def measure(
+    pin_name: str,
+    voltage_level: float,
+    voltage_level_range: float,
+    current_limit: float,
+    current_limit_range: float,
+    source_delay: float,
+) -> Tuple[float, float]:
+    """Source and measure a DC voltage with an NI SMU connected via an NI-SWITCH multiplexer."""
+    logging.info("Executing measurement: pin_name=%s voltage_level=%g", pin_name, voltage_level)
+
+    cancellation_event = threading.Event()
+    measurement_service.context.add_cancel_callback(cancellation_event.set)
+
+    with measurement_service.context.reserve_session(pin_name) as reservation:
+        with reservation.initialize_nidcpower_session(), reservation.initialize_niswitch_multiplexer_session():
+            # Configure the SMU channel connected to the input pin.
+            connection = reservation.get_nidcpower_connection_with_multiplexer(
+                niswitch.Session, pin_name
+            )
```

Comment:
Was there a specific reason for asking this question? In prior conversations, my feedback was I don't think we should be using "get connections" APIs and then reconstructing a channel list to ensure the channel list order matches the pin list order passed to the measurement. I would prefer the channel list returned from the session reservation meet this contract rather than requiring users to write extra code to meet this contract.

I'm not sure if the same thing is possible or not with switches. I assume we can still get 1 or N switch sessions through the initialization of the switch sessions the same way it is with instrument sessions. However, I don't know if there is any way to get a list of routes to apply with using "get connections" APIs. This example only uses a single pin so it's also not entirely clear what the code would look like if you need to apply N routes for N pins.

If nothing else, maybe the example should still use the session reservation info when programming the instrument session and then use "get connections" to get the multiplexer session/route info if that's the only way to do that?

Reply:
> Was there a specific reason for asking this question?

Just a request for feedback

> I'm not sure if the same thing is possible or not with switches. I assume we can still get 1 or N switch sessions through the initialization of the switch sessions the same way it is with instrument sessions. However, I don't know if there is any way to get a list of routes to apply with using "get connections" APIs. 

I don't think it's possible.

> This example only uses a single pin so it's also not entirely clear what the code would look like if you need to apply N routes for N pins.

If all the pins are handled the same, use get_connections_..._with_multiplexer.

If some of the pins are different, either look at the per-connection pin name or call get_connection_..._with_multiplexer N times.

> If nothing else, maybe the example should still use the session reservation info when programming the instrument session and then use "get connections" to get the multiplexer session/route info if that's the only way to do that?

That seems reasonable to me.

Reply:
I forget the exact syntax for `connect_multiple`, but I think it accepts a comma separated list of routes. If the `TypedMultiplexerSessionInformation` returned a comma separated route list (just like `TypedSessionInformation` has a `channel_list`), then you wouldn't need to use the "get connections" APIs at all. 

I don't really have any practical experience with programming switches so it's difficult for me to say whether it would be more common/desirable to use `connect_multiple` with a route list in this fashion vs. calling the "get connections" and `connect_multiple` APIs in a loop.

Reply:
I think merging multiplexer connections would introduce several new assumptions.
- `multiplexer_route` is in a user-defined format, and it may or may not make sense to separate it with commas or semicolons or whatever. Standard multiplexer type ids might help identify whether this is a safe assumption.
- All pins for the instrument session are using the same multiplexer session.
- All pins for the instrument session are allowed to be connected at the same time. This is not a given. Some measurements may need to iterate over pins and connect one at a time.

So I think it makes sense at this time to use `session_information.channel_list` to program the instrument session and use `get_..._connection_with_multiplexer` to program the multiplexer session(s).